### ISSUE TITLE
New `fastavro` release breaks `alert_stream` example

### ISSUE THREAD

**Issue by stefanv** [lsst-dm/alert_stream#24]
The Dockerfile doesn't currently pin `fastavro`.  Using version <= 0.21.23 works, whereas versions >= 0.21.24 give the following error:

<details>

```
$ docker run -it --rm \    
     --network=alertstream_default \
     -v $PWD/data:/home/alert_stream/data:ro \
     alert_stream python bin/sendAlertStream.py kafka:9092 my-stream
visit: 1574. 	time: 1560882570.0124252
Task exception was never retrieved
future: <Task finished coro=<schedule_delays() done, defined at bin/sendAlertStream.py:30> exception=RuntimeError('failed to find alert data in ./data/alerts_11574.avro',)>
Traceback (most recent call last):
  File "/home/sample-avro-alert/python/lsst/alert/packet/io.py", line 46, in retrieve_alerts
    reader = fastavro.reader(fp, reader_schema=reader_schema.definition if reader_schema else None)
  File "fastavro/_read.pyx", line 745, in fastavro._read.reader.__init__
  File "fastavro/_read.pyx", line 717, in fastavro._read.file_reader.__init__
  File "fastavro/_schema.pyx", line 55, in fastavro._schema.parse_schema
  File "fastavro/_schema.pyx", line 167, in fastavro._schema._parse_schema
  File "fastavro/_schema.pyx", line 204, in fastavro._schema.parse_field
  File "fastavro/_schema.pyx", line 66, in fastavro._schema._parse_schema
  File "fastavro/_schema.pyx", line 112, in fastavro._schema._parse_schema
  File "fastavro/_schema.pyx", line 157, in fastavro._schema._parse_schema
fastavro._schema_common.SchemaParseException: redefined named type: lsst.alert.diaSource

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "bin/sendAlertStream.py", line 47, in schedule_delays
    yield from asyncio.ensure_future(delay(wait_time, function, arg))
  File "bin/sendAlertStream.py", line 27, in delay
    return function(*args)
  File "bin/sendAlertStream.py", line 76, in send_visit
    schema, alert_packets = retrieve_alerts(file_data)
  File "/home/sample-avro-alert/python/lsst/alert/packet/io.py", line 49, in retrieve_alerts
    f"{fp.name if hasattr(fp, 'name') else 'stream'}") from e
RuntimeError: failed to find alert data in ./data/alerts_11574.avro
```

</details>

I suspect the change that caused this is:

```
commit 589493942520b447e06f1989a97f0ae9797ba99c
Author: scottbelden <scottabelden@gmail.com>
Date:   Sun May 12 16:24:41 2019 -0400

    ensure that named schemas are not redefined
```

Did fastavro become more strict, and did alert_stream override a schema definition before?  Any thoughts, @scottbelden?

**Comment by scottbelden** [comment]
Yes, this change would have caused fastavro to be more strict about the schema definition of named types. You mentioned version `0.21.23` works (as it wasn't as strict), so would you be able to use that version to print the schema here so I could take a look at it? Or if you have the avro file and can post that somewhere I could get the schema from it.

**Comment by jdswinbank** [comment]
Confirmed that I see the same behaviour here.

I'll pin the version of fastavro being used on [DM-20232](https://jira.lsstcorp.org/browse/DM-20232), which should help you avoid the immediate problem.

Longer term, we should fix this data — it has a fairly murky origin, it wasn't created using a modern version of our schema, and could generally use an overhaul. I'll add that to the backlog, but won't get to it today.

@scottbelden, thanks very much for your interest in this! Should you want to take this further, the data is available from https://github.com/lsst-dm/alert_stream/tree/master/data. However, per the above, I'm not altogether surprised that there are issues with this data, and we should be able to resolve it internally.

**Comment by jdswinbank** [comment]
Pin added. Data will be upgraded on [DM-17551](https://jira.lsstcorp.org/browse/DM-17551) (but likely not today, sorry).

**Comment by scottbelden** [comment]
For that particular schema, `lsst.alert.diaSource` is defined three times, `lsst.alert.diaObject` is defined twice, and `lsst.alert.cutout` is defined twice. For example, in the case of `cutout`, it currently looks like this:

```
  {'default': None,
   'name': 'cutoutDifference',
   'type': [{'type': 'record',
     'name': 'lsst.alert.cutout',
     'fields': [{'name': 'fileName', 'type': 'string'},
      {'name': 'stampData', 'type': 'bytes'}]},
    'null']},
  {'default': None,
   'name': 'cutoutTemplate',
   'type': [{'type': 'record',
     'name': 'lsst.alert.cutout',
     'fields': [{'name': 'fileName', 'type': 'string'},
      {'name': 'stampData', 'type': 'bytes'}]},
    'null']}
```

But it should look like this:

```
{'default': None,
   'name': 'cutoutDifference',
   'type': [{'type': 'record',
     'name': 'lsst.alert.cutout',
     'fields': [{'name': 'fileName', 'type': 'string'},
      {'name': 'stampData', 'type': 'bytes'}]},
    'null']},
  {'default': None,
   'name': 'cutoutTemplate',
   'type': ['lsst.alert.cutout',
    'null']}
```

Hope that helps!

**Comment by stefanv** [comment]
Thanks for the helpful feedback @scottbelden, and thank you for the quick fix @jdswinbank!

**Comment by jdswinbank** [comment]
Thank you very much @scottbelden!

**Comment by ebellm** [comment]
Created [DM-23606](https://jira.lsstcorp.org/browse/DM-23606) to resolve the duplication of nested schemas that is the root cause of this issue.

**Comment by jdswinbank** [comment]
DM-23606 has been resolved (by #28, and https://github.com/lsst-dm/sample-avro-alert/pull/8) which closes this issue. Be aware that further schema updates will be forthcoming.

**Comment by scottbelden** [comment]
I'm glad this could be resolved. 👍
